packages/node-resolve/src/resolveId.js

```diff
@@ -0,0 +1,130 @@
+import fs from 'fs';
+import path from 'path';
+import { promisify } from 'util';
+
+import resolve from 'resolve';
+
+import { getPackageName } from './util';
+
+const resolveImportPath = promisify(resolve);
+const readFile = promisify(fs.readFile);
+
+const pathNotFoundError = (subPath, pkgPath) =>
+  new Error(`Package subpath '${subPath}' is not defined by "exports" in ${pkgPath}`);
+
+function findExportKeyMatch(keys, subPath) {
+  return keys.find((key) => (key.endsWith('/') ? subPath.startsWith(key) : key === subPath));
+}
+
+function mapSubPath(pkgJsonPath, subPath, key, value) {
+  if (typeof value === 'string') {
+    if (value.endsWith('/')) {
+      return `${value}${subPath.substring(key.length)}`;
+    }
+    // mapping is a string, for example { "./foo": "./dist/foo.js" }
+    return value;
+  }
+
+  if (Array.isArray(value)) {
+    // mapping is an array with fallbacks, for example { "./foo": ["foo:bar", "./dist/foo.js"] }
+    return value.find((v) => v.startsWith('./'));
+  }
+
+  throw pathNotFoundError(subPath, pkgJsonPath);
+}
+
+function findEntrypoint(pkgJsonPath, subPath, exportMap, conditions, key) {
+  if (typeof exportMap !== 'object') {
+    return mapSubPath(pkgJsonPath, subPath, key, exportMap);
+  }
+
+  // iterate conditions recursively, find the first that matches all conditions
+  for (const [condition, subExportMap] of Object.entries(exportMap)) {
+    if (conditions.includes(condition)) {
+      const mappedSubPath = findEntrypoint(pkgJsonPath, subPath, subExportMap, conditions, key);
+      if (mappedSubPath) {
+        return mappedSubPath;
```

Comment:
Note that condition matching will actually fall back if not found.

Eg:

```json
{
  "exports": {
    "node": {
      "require": "./cjs.js"
    },
    "default": "./main.js"
  }
}
```

will use the default in Node.js when using `import`.

Reply:
We handle this case, I added a test to confirm :)